### DISCUSSION TITLE
How can I debug freezes in LTeX?

### DISCUSSION THREAD

**Discussion by tobiasdiez** [valentjn/vscode-ltex#132] [Q&A] [answered]
From time to time, LTeX has a constantly high CPU usage and I have to manually kill the executable. Sadly, I haven't found a way to reproduce it.

Can I do something to narrow down the origin of the bug? How do you debug or get memory dumbs from extensions?

**Version information**

- Operating system: Windows 10
- VS Code: 1.49 (but occurred also in earlier releases)
- vscode-ltex: 7.2 (but occurred also in earlier releases)
- ltex-ls: automatically downloaded
- Java: I have JDK 14 and 15 installed, but it could also be the automatically downloaded version

**Comment by valentjn** [comment, marked as answer]
Read the docs on [how to report bugs](https://valentjn.github.io/vscode-ltex/docs/contributing-code-issues.html#how-to-report-bugs). It's also a little bit about how to debug. Especially the logs can help to understand the issue.

The no. 1 suspect for freezes like this is wrong settings caching like in https://github.com/valentjn/vscode-ltex/issues/29#issuecomment-622381226. Probably your bug is a duplicate of #96 (unfortunately, the submitter did not provide sufficient info to help me find the bug). In 7.1.2, I added specific logging to be able to better debug this (as that bug is still in there). Look for lines containing `Reinitializing LanguageTool due to different settings` in the server log.

Even if it's not due to wrong settings caching, looking at the logs still might help. If you want help, please submit a new complete bug report; use the command [`LTeX: Report bug in LTeX`](https://valentjn.github.io/vscode-ltex/docs/commands.html#ltex-report-bug-in-ltex) as it fills out all fields for you.

Regarding the more general question of how to debug VS Code extensions, the canonical way is to clone the repo, run `npm i`, open the repo in VS Code, and press `F5`. However, [since the bug is almost surely in the Java part ltex-ls and vscode-ltex basically only starts ltex-ls](https://valentjn.github.io/vscode-ltex/docs/faq.html#whats-the-difference-between-vscode-ltex-ltex-ls-and-languagetool), that won't help here. Instead, the naive approach is to clone ltex-ls and put additional debug logging statements in there (you don't even need to clone vscode-ltex then, just use [`ltex.ltex-ls.path`](https://valentjn.github.io/vscode-ltex/docs/settings.html#ltexltex-lspath)). I haven't looked for a way to have vscode-ltex start Java in some kind of debug mode with breakpoints etc., but that would of course be better.

**Comment by tobiasdiez** [comment]
Thanks for the information, and sorry for not being able to provide more information. The bug that I'm experiencing is somewhat different in that the CPU usage is at about 20% (all in the ltex-ls java process) but it hangs there constantly. That is, even after waiting for several mins nothing happens and I have to kill the exe myself. I'll try to investigate a bit more the next time the bug occurs, and will open a new issue then.
